contracts/src/TeleporterTokenSource.sol

```diff
@@ -134,30 +192,67 @@ abstract contract TeleporterTokenSource is ITeleporterTokenBridge, TeleporterOwn
         bytes32 sourceBlockchainID,
         address originSenderAddress,
         bytes memory message
-    ) internal virtual override {
-        (SendTokensInput memory input, uint256 amount) =
-            abi.decode(message, (SendTokensInput, uint256));
+    ) internal override {
+        BridgeMessage memory bridgeMessage = abi.decode(message, (BridgeMessage));
 
         // Check that bridge instance returning has sufficient amount in balance
         uint256 senderBalance = bridgedBalances[sourceBlockchainID][originSenderAddress];
-        require(senderBalance >= amount, "TeleporterTokenSource: insufficient bridge balance");
+        require(
+            senderBalance >= bridgeMessage.amount,
+            "TeleporterTokenSource: insufficient bridge balance"
+        );
 
         // Decrement the bridge balance by the unwrap amount
-        bridgedBalances[sourceBlockchainID][originSenderAddress] = senderBalance - amount;
-
-        // If the final destination of the transfer is this source bridge instance, withdraw the tokens.abi
-        // Otherwise, perform a multihop transfer to the input destination bridge instance.
-        if (input.destinationBlockchainID == blockchainID) {
-            require(
-                input.destinationBridgeAddress == address(this),
-                "TeleporterTokenSource: invalid bridge address"
+        bridgedBalances[sourceBlockchainID][originSenderAddress] =
+            senderBalance - bridgeMessage.amount;
+
+        if (bridgeMessage.messageType == BridgeMessageType.SINGLE_HOP_SEND) {
+            SingleHopSendMessage memory payload =
+                abi.decode(bridgeMessage.payload, (SingleHopSendMessage));
+            emit TokensWithdrawn(payload.recipient, bridgeMessage.amount);
+            _withdraw(payload.recipient, bridgeMessage.amount);
+            return;
+        } else if (bridgeMessage.messageType == BridgeMessageType.SINGLE_HOP_CALL) {
+            SingleHopCallMessage memory payload =
+                abi.decode(bridgeMessage.payload, (SingleHopCallMessage));
+            _handleSendAndCall(payload, bridgeMessage.amount);
+            return;
+        } else if (bridgeMessage.messageType == BridgeMessageType.MULTI_HOP_SEND) {
+            MultiHopSendMessage memory payload =
+                abi.decode(bridgeMessage.payload, (MultiHopSendMessage));
+            _send(
+                SendTokensInput({
+                    destinationBlockchainID: payload.destinationBlockchainID,
+                    destinationBridgeAddress: payload.destinationBridgeAddress,
+                    recipient: payload.recipient,
+                    primaryFee: payload.secondaryFee,
+                    secondaryFee: 0,
+                    requiredGasLimit: payload.secondaryGasLimit
+                }),
+                bridgeMessage.amount,
+                true
+            );
+            return;
+        } else if (bridgeMessage.messageType == BridgeMessageType.MULTI_HOP_CALL) {
```

Comment:
I originally had a catch all `else` block here, but didn't think it was necessary given that we enumerate all of the `BridgeMessageType` values already (which made it difficult/impossible to reach that case with a unit test). In `TeleporterTokenDestination`, there are `BridgeMessageType` values that are invalid for it to receive, so it's a possible invalid case.

Reply:
I was worried about the case where someone makes a destination bridge contract and they are encoding the message back incorrectly, but I guess in this case when we `abi.decode` to a BridgeMessageType it would revert.